refs-and-the-dom guide translation (#122)
reactjs/de.reactjs.org: content/docs/refs-and-the-dom.md

permalink: docs/refs-and-the-dom.html
---

> Refs bieten eine Möglichkeit, auf DOM-Knotens oder React-Elemente zuzugreifen, die mit der Rendermethode erstellt wurden.

Reviewer: ```suggestion
Refs bieten eine Möglichkeit, auf DOM-Knoten oder React-Elemente zuzugreifen, die mit der Rendermethode erstellt wurden.
```

> Im typischen React-Datenfluss sind [Props](/docs/components-and-props.html) die einzige Möglichkeit, wie übergeordnete Komponenten mit ihren untergeordneten Komponenten interagieren. Um ein untergeordnetes Element zu ändern, rendere es mit neuen Props neu. Es gibt jedoch einige Fälle, in denen du ein untergeordnetes Element unbedingt außerhalb des typischen Datenflusses ändern musst. Das zu ändernde untergeordnete Element kann eine Instanz einer React-Komponente oder ein DOM-Element sein. In beiden Fällen bietet React eine Notluke.

Reviewer: ```suggestion
Im normalen React-Datenfluss stellen [Props](/docs/components-and-props.html) die einzige Möglichkeit dar, um übergeordnete Komponenten mit ihren untergeordneten Komponenten interagieren zu lassen. Um ein untergeordnetes Element zu verändern, muss es mit neuen Props erneut gerendert werden. Es gibt jedoch einige Fälle, in denen du ein untergeordnetes Element unbedingt außerhalb des typischen Datenflusses ändern musst. Das zu ändernde untergeordnete Element kann eine Instanz einer React-Komponente oder ein DOM-Element sein. In beiden Fällen bietet React einen Ausweg.
```

### Wann man Refs benutzt {#when-to-use-refs}

Es gibt einige gute Anwendungsfälle für Refs:

* Verwalten des Fokus, der Textauswahl oder der Medienwiedergabe.
* Imperative Animationen auslösen.
* Integration in DOM-Bibliotheken von Drittanbietern.

Vermeide die Verwendung von Refs für alles, was deklarativ durchgeführt werden kann.

Anstatt beispielsweise die Methoden `open ()` und `close ()` in einer `Dialog` -Komponente verfügbar zu machen, übergebe ihr eine` isOpen`-Prop.

> ### Überbeanspruche Refs nicht {#dont-overuse-refs}

Reviewer: ```suggestion
### Verwende Refs nicht zu häufig {#dont-overuse-refs}
```

> Deine erste Neigung könnte sein, Refs zu verwenden, um Dinge in deier App "geschehen zu lassen". Wenn dies der Fall ist, nehme dich einen Moment Zeit und überlege genauer, wo sich der Status in der Komponentenhierarchie befinden sollte. Oft wird klar, dass der richtige Ort, um diesen Zustand zu "besitzen", auf einer höheren Ebene in der Hierarchie liegt. Beispiele hierfür findest du im Handbuch [State anheben](/docs/lifting-state-up.html).

Reviewer: ```suggestion
Deine erste Intuition könnte sein, refs zu verwenden, um Dinge in deiner App "geschehen zu lassen". Wenn dies der Fall ist, nehme dir einen Moment Zeit und überlege genauer, wo sich der Status in der Komponentenhierarchie befinden sollte. Oft wird klar, dass der richtige Ort, um diesen Zustand zu "besitzen", auf einer höheren Ebene in der Hierarchie liegt. Beispiele hierfür findest du im Handbuch [State anheben](/docs/lifting-state-up.html).
```

Reviewer: Not happy with `um diesen Zustand zu "besitzen"`, but couldn't come up with a better alternative for "besitzen".

Author: Why make Refs lowercase?

Reviewer 2: @jakeherp @wogsland I have changed it so that it can be merged.

Du könntest zwar [Hinzufügen eines Ref zu einer Klassenkomponente](#adding-a-ref-to-a-class-component), dies ist jedoch keine ideale Lösung, da du nur eine Komponenteninstanz anstelle eines DOM-Knotens erhalten würdest. Darüber hinaus würde dies nicht mit Funktionskomponenten funktionieren.

Wenn du React 16.3 oder höher verwendest, empfehlen wir in diesen Fällen die Verwendung von [Weiterleiten von Refs](/docs/forwarding-refs.html). **Mit der Weiterleitung von Refs können Komponenten festlegen, dass der Ref jeder untergeordneten Komponente als ihr eigener angezeigt wird**. Ein detailliertes Beispiel, wie der DOM-Knoten eine untergeordnete Komponente für eine übergeordnete Komponente verfügbar gemacht wird, findest du [in der Dokumentation zur Weiterleiten von Refs](/docs/forwarding-refs.html#forwarding-refs-to-dom-components).

Wenn du React 16.2 oder eine niedrigere Version verwendest oder mehr Flexibilität benötigst, als durch die Weiterleiten von Refs bereitgestellt wird, kannst du [diesen alternativen Ansatz](https://gist.github.com/gaearon/1a018a023347fe1c2476073330cc5509) verwenden und einen Ref explizit als übergeben anders benanntes Prop.

> Wenn möglich, raten wir davon ab, DOM-Knoten freizulegen, dies kann jedoch eine nützliche Notluke sein. Beachte, dass du bei diesem Ansatz der untergeordneten Komponente Code hinzufügen musst. Wenn du absolut keine Kontrolle über die Implementierung der untergeordneten Komponente hast, kannst du als letzte Option [`findDOMNode ()`](/docs/react-dom.html#finddomnode) verwenden, dies wird jedoch in [`StrictMode`](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage) nicht empfohlen und ist veraltet.

Reviewer: ```suggestion
Wenn möglich, raten wir davon ab, DOM-Knoten freizulegen, dies kann jedoch ein nützlicher Ausweg sein. Beachte, dass du bei diesem Ansatz der untergeordneten Komponente Code hinzufügen musst. Wenn du absolut keine Kontrolle über die Implementierung der untergeordneten Komponente hast, kannst du als letzte Option [`findDOMNode ()`](/docs/react-dom.html#finddomnode) verwenden, dies wird jedoch in [`StrictMode`](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage) nicht empfohlen und ist veraltet.
```
